network/p2p/connection.go

```diff
@@ -82,32 +87,166 @@ func (conn *connection) send(message *transport.NetworkMessage) error {
 	return nil
 }
 
-// sendMessages (blocking) reads messages from its outMessages channel and sends them to the peer until the channel is closed.
-func (conn connection) sendMessages() {
-	for message := range conn.outMessages {
-		if conn.messenger.Send(message) != nil {
-			log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+func (conn *managedConnection) exchange(receivedMessages messageQueue) {
+	conn.mux.Lock()
```

Comment:
otherwise access to `conn.outMessages` (and fields that `receiveMessages` accesses) is racy (according to `-race`).

E.g. calls to `close()` (called async from this func) nill those fields, thus theoretically causing racy behaviour (in practce not, but OK).